ForwardingDecision.cs

```diff
@@ -0,0 +1,79 @@
+/*
+Pax : tool support for prototyping packet processors
+Nik Sultana, Cambridge University Computer Lab, June 2016
+
+Use of this source code is governed by the Apache 2.0 license; see LICENSE.
+*/
+
+namespace Pax {
+
+  /// <summary>
+  /// A packet processor makes some sort of forwarding decision, in addition
+  /// to possibly analysing/modifying the packet (and generating new packets).
+  /// </summary>
+  /// <remark>
+  /// The forwarding decision makes clear whether the processor recommends to
+  /// * drop the packet (i.e., nothing more is to be done with it)
+  /// * forward to a single port
+  /// * forward to multiple ports
+  ///
+  /// NOTE this idiom simulates using algebraic types in C#. It's values are
+  ///      analysed using the Visitor pattern.
+  /// </remark>
+  /// <remark>
+  /// Note that there is a hierarchy between forwarding decisions: Drop is
+  /// subsumed by SinglePortForward (which allows forwarding to at most one
+  /// port), which is subsumed by MultiPortForward (which allows forwarding to
+  /// any number of ports, including one port, or none).
+  /// </remark>
+  public abstract class ForwardingDecision {
+    private ForwardingDecision() {}
+
+    // FIXME can use singleton to avoid multiple allocations of this?
+    /// <summary>
+    /// The decision to drop the packet.
+    /// </summary>
+    public sealed class Drop : ForwardingDecision {
+      // We only need to keep a single instance of Drop, so we use a singleton pattern.
+      public static readonly Drop inst;
+      private Drop () {}
+      static Drop () {
+        Drop.inst = new Drop();
+      }
+      public static Drop instance() {
```

Comment:
Could replace the field, static constructor and method with a static property. E.g. in C# 6 syntax:
```C#
public static Drop Instance { get; } = new Drop();
```